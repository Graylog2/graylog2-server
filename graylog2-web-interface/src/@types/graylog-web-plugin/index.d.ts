/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import FetchError from 'logic/errors/FetchError';
import { NodeInfo } from 'stores/nodes/NodesStore';

interface PluginRoute {
  path: string;
  component: React.ComponentType;
  parentComponent?: React.ComponentType | null;
  permissions?: string;
}
interface PluginNavigation {
  path: string;
  description: string;
}
interface PluginNavigationItems {
  key: string;
  component: React.ComponentType<{ smallScreen?: boolean }>;
}
interface GlobalNotification {
  key: string;
  component: React.ComponentType;
}

interface PluginPages {
  search?: {
    component: React.ComponentType;
  }
}

interface PluginPageFooter {
  component: React.ComponentType;
}

interface PluginForwarder {
  ForwarderReceivedBy: React.ComponentType<{
    inputId: string;
    forwarderNodeId: string;
  }>;
  isLocalNode: (nodeId: string) => NodeInfo;
  messageLoaders: {
    ForwarderInputDropdown: React.ComponentType<{
      autoLoadMessage?: boolean;
      preselectedInputId?: string;
      title?: string;
      label?: string;
      loadButtonDisabled?: boolean;
      onLoadMessage: (selectedInput: string) => void;
    }>;
  };
}

interface PluginCloud {
  oktaUserForm: {
    fields: {
      username: React.ComponentType<{}> | null;
      email: React.ComponentType<{}>;
      password: React.ComponentType<{}>;
    };
    validations: {
      password: (errors: { [name: string]: string }, password: string, passwordRepeat: string) => { [name: string]: string };
    };
    extractSubmitError: (errors: FetchError) => string;
    onCreate: (formData: { [name: string ]: string }) => { [name: string]: string };
  };
}
interface InputConfiguration {
  type: string;
  component: React.ComponentType<{}>;
  embeddedComponent?: React.ComponentType<{}>;
}
declare module 'graylog-web-plugin/plugin' {
  interface PluginExports {
    navigation?: Array<PluginNavigation>;
    navigationItems?: Array<PluginNavigationItems>;
    globalNotifications?: Array<GlobalNotification>
    routes?: Array<PluginRoute>;
    pages?: PluginPages;
    pageFooter?: Array<PluginPageFooter>;
    cloud?: Array<PluginCloud>;
    forwarder?: Array<PluginForwarder>;
    inputConfiguration?: Array<InputConfiguration>
  }

  interface PluginRegistration {
    exports: PluginExports;
  }

  interface PluginManifest extends PluginRegistration {
    new (json: {}, exports: PluginExports): PluginManifest;
  }

  interface PluginStore {
    register: (manifest: PluginRegistration) => void;
    exports: <T extends keyof PluginExports>(key: T) => PluginExports[T];
    unregister: (manifest: PluginRegistration) => void;
  }

  const PluginStore: PluginStore;
  const PluginManifest: PluginManifest;
}
