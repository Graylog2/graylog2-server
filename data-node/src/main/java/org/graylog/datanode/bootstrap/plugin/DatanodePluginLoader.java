/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
package org.graylog.datanode.bootstrap.plugin;

import com.google.common.collect.ImmutableSet;
import org.graylog2.plugin.Plugin;
import org.graylog2.plugin.PluginBootstrapConfig;
import org.graylog2.shared.plugins.ChainingClassLoader;
import org.graylog2.shared.plugins.PluginLoader;

import java.io.File;
import java.util.ServiceLoader;
import java.util.Set;

public class DatanodePluginLoader extends PluginLoader {

    public DatanodePluginLoader(File pluginDir, ChainingClassLoader classLoader) {
        super(pluginDir, classLoader);
    }

    @Override
    @SuppressWarnings("unchecked")
    public Set<PluginBootstrapConfig> loadPluginBootstrapConfigs() {
        return ImmutableSet.copyOf(
                (ServiceLoader<PluginBootstrapConfig>) (ServiceLoader<?>)
                        ServiceLoader.load(DatanodePluginBootstrapConfig.class, classLoader)
        );
    }

    @Override
    @SuppressWarnings("unchecked")
    protected Iterable<Plugin> loadClassPathPlugins() {
        return (ServiceLoader<Plugin>) (ServiceLoader<?>)
                ServiceLoader.load(DatanodePlugin.class);
    }

    @Override
    @SuppressWarnings("unchecked")
    protected Iterable<Plugin> loadJarPlugins() {
        return ImmutableSet.copyOf(
                (ServiceLoader<Plugin>) (ServiceLoader<?>)
                        ServiceLoader.load(DatanodePlugin.class, classLoader)
        );
    }

}
