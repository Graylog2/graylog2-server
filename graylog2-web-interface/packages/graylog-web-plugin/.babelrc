{
  "presets": ["env", "stage-0"],
  "plugins": ["add-module-exports"]
}
