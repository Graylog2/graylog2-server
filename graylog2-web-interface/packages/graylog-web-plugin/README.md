Graylog Web Plugin Archetype
============================

This module is supposed to help with developing web interface plugins for [Graylog](http://www.graylog.org). It provides classes which help generating a webpack configuration, defining the plugin manifest (metadata as well as entities which are provided to the web interface) as well as the `PluginStore` class which is used for registering plugins to make them accessible for the web interface.

This module is supposed to be used in conjunction with the maven archetype which can be found [here](https://github.com/Graylog2/graylog-plugin-archetype) and requires a checkout of the [Graylog server repository](https://github.com/Graylog2/graylog2-server) in some place.

## Installation

`npm install --save graylog-web-plugin`

## Usage

```
$ npm init
[...]
$ npm --save-dev graylog-web-plugin
```

Create a file named `build.config.js` in your plugin directory, looking like this:

```
module.exports = {
  web_src_path: '<path to your Graylog git repository>',
};
```

and use a webpack.config.js like this: 

```
const PluginWebpackConfig = require('graylog-web-plugin').PluginWebpackConfig;
const loadBuildConfig = require('graylog-web-plugin').loadBuildConfig;
const path = require('path');
const buildConfig = loadBuildConfig(path.resolve(__dirname, './build.config'));

module.exports = new PluginWebpackConfig('my.fully.qualified.plugin.classname', buildConfig, {
  // Here goes your additional webpack configuration.
});
```

The third (optional) parameter for the `PluginWebpackConfig` constructor is an object which is merged into the generated webpack config, so you can add/overwrite any part of the generated configuration.

## Contributing

Feel free to contribute to this collection of helpers by forking the repository and submitting a pull request. Thanks!

## Release a new version

    $ npm version [<newversion> | major | minor | patch]
    $ git push origin master && git push --tags
    $ npm publish

## Release History
* 0.0.21 Add optimizations for production builds.
* 0.0.20 Add plugin fqdn as prefix of built `module.json`.
* 0.0.19 Add webpack config to compile typescript code in plugins.
* 0.0.18 Removing now unneeded shared bundle.
* 0.0.17 Bugfix, using absolute filename for build config.
* 0.0.16 Adding build config file including option specifying location of Graylog server repo checkout.
* 0.0.15 Bugfix, add missing import
* 0.0.14 Add shared and vendor bundle to generated webpack config
* 0.0.13 First actually usable version
