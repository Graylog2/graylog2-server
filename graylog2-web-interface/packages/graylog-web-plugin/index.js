module.exports = {
  PluginWebpackConfig: require('./lib/PluginWebpackConfig'),
  loadBuildConfig: require('./lib/loadBuildConfig'),
};
