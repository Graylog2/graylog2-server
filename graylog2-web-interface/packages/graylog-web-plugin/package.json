{
  "name": "graylog-web-plugin",
  "version": "0.0.21",
  "description": "Helper code for streamlining Graylog web interface plugin development",
  "main": "index.js",
  "scripts": {
    "test": "eslint src/*",
    "compile": "babel -d lib/ src/",
    "prepublish": "npm run compile"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/Graylog2/graylog-web-plugin.git"
  },
  "keywords": [
    "graylog",
    "web",
    "plugin"
  ],
  "author": "Graylog, Inc. <hello@graylog.com>",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/Graylog2/graylog-web-plugin/issues"
  },
  "homepage": "https://github.com/Graylog2/graylog-web-plugin#readme",
  "dependencies": {
    "babel-core": "^5.8.25",
    "babel-loader": "^5.3.2",
    "html-webpack-plugin": "^1.7.0",
    "webpack": "^1.12.9",
    "webpack-cleanup-plugin": "^0.1.1",
    "webpack-merge": "^0.7.1"
  },
  "devDependencies": {
    "babel": "^5.8.34",
    "babel-eslint": "^6.1.0",
    "eslint": "^2.13.1",
    "eslint-config-graylog": "^1.0.2"
  }
}
