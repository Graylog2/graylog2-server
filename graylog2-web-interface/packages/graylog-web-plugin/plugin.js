module.exports = {
  PluginManifest: require('./lib/PluginManifest'),
  PluginStore: require('./lib/PluginStore'),
}
