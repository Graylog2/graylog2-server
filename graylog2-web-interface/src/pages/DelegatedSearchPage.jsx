import React from 'react';

import { PluginStore } from 'graylog-web-plugin/plugin';
import SearchPage from './SearchPage';

export default (props) => {
  const components = PluginStore.exports('pages')
    .map(c => c.search || {})
    .map(c => c.component)
    .filter(c => c) || [];
  const Component = components[0] || SearchPage;
  return <Component {...props} />;
};
