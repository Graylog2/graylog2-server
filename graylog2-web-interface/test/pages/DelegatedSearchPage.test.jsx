jest.mock('c3', () => ({ default: jest.fn() }))
  .mock('components/search/LegacyHistogram', () => ({ default: jest.fn() }))
  .mock('components/search/QueryInput', () => ({ default: jest.fn() }))
  .mock('injection/CombinedProvider', () => new (require('helpers/mocking').CombinedProviderMock)());

import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { PluginManifest, PluginStore } from 'graylog-web-plugin/plugin';

import DelegatedSearchPage from 'pages/DelegatedSearchPage';

test('Renders SearchPage by default', () => {
  const renderer = new ShallowRenderer();
  const tree = renderer.render(
    <DelegatedSearchPage />
  );
  expect(tree).toMatchSnapshot();
});

test('Renders other component if registered', () => {
  const renderer = new ShallowRenderer();
  const SimpleComponent = () => <div>Hello!</div>;

  PluginStore.register(new PluginManifest({}, {
    pages: {
      search: { component: SimpleComponent },
    },
  }));

  const tree = renderer.render(
    <DelegatedSearchPage />
  );

  expect(tree).toMatchSnapshot();
});
