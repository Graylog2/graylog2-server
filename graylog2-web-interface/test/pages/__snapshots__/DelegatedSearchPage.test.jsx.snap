// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders SearchPage by default 1`] = `<SearchPage />`;

exports[`Renders other component if registered 1`] = `<SimpleComponent />`;
